cEP-0028.md

```diff
@@ -0,0 +1,150 @@
+# Optimizing coala for language server
+
+| Metadata |                                       |
+| -------- | ------------------------------------- |
+| cEP      | 0028                                  |
+| Version  | 0.1                                   |
+| Title    | Optimizing coala for language server  |
+| Authors  | Kilari Teja <mailto:ksdme9@gmail.com> |
+| Status   | Proposed                              |
+| Type     | Process                               |
+
+## Abstract
+
+This cEP describes the optimizations that need to be carried out
+to improve the performance of the coala based language server.
+
+## Introduction
+
+Language servers are the tools used to serve the code editors with
+functionalities such as linting, code fixing, analysis etc in a
+standalone and independent manner. coala as a code style checker,
+linting tool and code fixer could be a powerful tool to be integrated
+into the code editor.
+
+Code Editing is a continuous process and the editors are expected
+to provide feedback on the fly, coala on the other hand was primarily
+designed as an analysis tool to be used after an incremental update to
+the codebase. Thus, the current implementation of a coala language
+server is severely limited by its performance.
+
+The performance of the server can be improved by primarily optimizing
+the coala and language server interface and improving the bear configuration
+options to avoid long running bears.
+
+## Proposal
+
+The optimization process focuses to maintain minimal changes to coala core
+and rather optimize the coala and language server interface. The following
+are the proposed ideas.
+
+1. Extending coala configuration parameters
+   Some bears are not often required to be executed on every cycle of coala
+   analysis especially during use cases such as this. The parameter set can
+   thus be extended to support parameters of type `cls_*`. Since these
+   parameters are not associated with any particular bear but rather control
+   the effect of various sections during executing coala, no support extension
+   for bears or coala-core would be necessary. These parameters enable support
+   for actions such as enabling/disabling a section for coala language server
+   or hooking a section for a particular request type. For example:
+
+   ```
+    [longrunning]
+    enabled = True
+    cls_enabled = False
+    bears = SomeLongRunningBear
+   ```
+
+   This kind of a configuration will disable the bear completely during
+   analysis cycles of coala via a language server while enabling it for
+   regular invocations.
+
+   ```
+    [longrunning]
+    enabled = False
+    cls_on = save
+    files = *.py
+    bears = SomeLongRunningBear
+   ```
+
+   This example shows a configuration where the section is disabled for coala
+   but will only be enabled when the code editor dispatches a request of type
```

Comment:
On save / edit should be generic . All other editor plugins have the same requirements.

Note Im not keen on this approach, but it is definately a viable and reasonable approach, and possibly the most feasible in time available.

The main problem with it is that it centralises editor config into .coafile which is shared by all devs.

The underlying problem is the editor can already control each section with 'enabled = false' which can be set with cli args and should be customisable via API also.

So this proposed setting may not provide any additional long term value. I still need to think about it though.

Reply:
> On save / edit should be generic . All other editor plugins have the same requirements.

The proposal is to support selecting which sections to execute during `save` and `edit` events.


> The main problem with it is that it centralises editor config into .coafile which is shared by all devs.

This approach was chosen compared to the other options available because this would enforce minimal requirements on the editor plugins and minimal changes to coala core. But I agree with your concern here. The other viable options would be to:

- have a `.coalals` file with same syntax as that of `.coafile` so they can use the same underlying parsing API's. This file will always be tiny and reference sections from `.coafile`.
- Use `workspace/configuration` (https://github.com/Microsoft/language-server-protocol/blob/gh-pages/specification.md#configuration-request-arrow_right_hook) and push the configuring to the plugin's end. This could possibly make designing plugins complex and could cause some confusion because of referencing sections from `.coafile` in editor config. Note that this request type was added to spec recently in version 3.6.
- Provide a tool to clean the configuration(`.coafile`) file i.e. a tool that will strip all the config parameters for `cls`. The configuration can thus be preprocessed before publishing.

> The underlying problem is the editor can already control each section with 'enabled = false' which can be set with cli args and should be customisable via API also.

We will still need a way by which the user can configure sections on the basis of request types. I plan to use the API to enable and disable sections.

Reply:
All editors have save and edit events, typically exposed to editor plugins.

They dont use those events because coala cant be used that way yet.

Effectively all you are doing is tagging sections with labels which group them into user-actions.  This could be used in the cli even, to run all bears in group 'save' on save but more section on 'pre-commit', and 'ci' for to be run in CI and even 'appveyor' to only run a subset on appveyor. These labels can have predefined means which all editor plugins use, if they want to.

An implicit group that already exists is 'global' bears which are typically going to be very slow. Also external 'linter' executables are a subset. And generic bears written in python (no extra runtime). Expose this info in the bears json and coala-ls can load it and make smart decisions, and disable bears using settings or other args.

Also each run data should include how much time was spent on each bear. coala-ls can then rerun the best bears.

Also another aspect is inter file dependency. That is a typical complexity for performing static analysis of large code bases.
Most of our NonGlobal bears only consider a single file at a time.

Imo you need to be declaring nonglobal bears which load multiple files as out of scope for optimising in this project. Too messy.

Reply:
> Effectively all you are doing is tagging sections with labels which group them into user-actions. This could be used in the cli even, to run all bears in group 'save' on save but more section on 'pre-commit', and 'ci' for to be run in CI and even 'appveyor' to only run a subset on appveyor. These labels can have predefined means which all editor plugins use, if they want to.

Thanks for this suggestion. This is definitely a better and more general way to design the requirement here. One specific change designing it this way incurs is support for this has to come from coala-core and not coala-ls, but since it holds the potential be a very general, I think we can make an exception about not causing changes to coala-core for the purpose of a better coala-ls.

> An implicit group that already exists is 'global' bears which are typically going to be very slow. Also external 'linter' executables are a subset. And generic bears written in python (no extra runtime). Expose this info in the bears json and coala-ls can load it and make smart decisions, and disable bears using settings or other args.

Can you be more specific about which bears.json you are speaking of? Do you mean coala-ls should maintain a bears.json(this wouldn't be a very good idea) or does coala-core maintain one such file that I am unaware of?

> Also each run data should include how much time was spent on each bear. coala-ls can then rerun the best bears.

We could certainly collect stats but I think the user should be in total control of which bears to run and when. We already support `window/showMessageRequest` so upon abnormal behavior where time required to serve the request is very huge we could notify the user and ask to disable the bears for the request type, although we can get feedback from the message and perform automatic disabling I do not think coala-ls should change `.coafile`.

> Imo you need to be declaring nonglobal bears which load multiple files as out of scope for optimising in this project. Too messy.

The mentioned optimizations could still be used with such bears, but we cannot guarantee the correctness of diagnostics in a few cases.

Reply:
Of course not maintaining a separate copy.
You can use `coala --json --bears` or something like that.  There is a bug in this atm, but that can and should be fixed.

We have other tools which fetch metadata about bears by reading from the Python objects directly.
e.g. https://github.com/coala/coala-bears/blob/master/.ci/generate_bear_requirements.py